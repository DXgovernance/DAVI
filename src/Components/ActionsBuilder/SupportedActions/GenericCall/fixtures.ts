import {
  ApproveSendTokens,
  DecodedCall,
  SupportedAction,
} from 'Components/ActionsBuilder/types';
import ERC20ABI from 'contracts/ERC20.json';
import { BigNumber, utils } from 'ethers';

const ERC20Contract = new utils.Interface(ERC20ABI.abi);

export const genericDecodedCallMock: DecodedCall = {
  from: '0x0000000000000000000000000000000000000000',
  callType: SupportedAction.GENERIC_CALL,
  function: ERC20Contract.getFunction('increaseAllowance'),
  to: '0x3f943f38b2fbe1ee5daf0516cecfe4e0f8734351',
  value: BigNumber.from(0),
  args: {
    param1: BigNumber.from(10000),
    param2: BigNumber.from(10000),
    param3: '0x3f943f38b2fbe1ee5daf0516cecfe4e0f8734351',
  },
};

export const approvetokenSpendingMock: ApproveSendTokens = {
  amount: BigNumber.from(1000),
  token: '0x3f943f38b2fbe1ee5daf0516cecfe4e0f8734351',
};
