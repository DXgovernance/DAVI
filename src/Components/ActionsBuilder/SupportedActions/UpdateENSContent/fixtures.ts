import { BigNumber, utils } from 'ethers';
import { DecodedCall } from 'Components/ActionsBuilder/types';
import { SupportedAction } from 'Components/ActionsBuilder/types';
import ENSPublicResolver from 'contracts/ENSPublicResolver.json';

const ENSPublicResolverContract = new utils.Interface(ENSPublicResolver.abi);

export const mockDecodedCallUpdateENSContent: DecodedCall = {
  from: '0x0000000000000000000000000000000000000000',
  callType: SupportedAction.ENS_UPDATE_CONTENT,
  function: ENSPublicResolverContract.getFunction('setContenthash'),
  to: '0x3f943f38b2fbe1ee5daf0516cecfe4e0f8734351',
  value: BigNumber.from(0),
  args: {
    node: '0x0000000000000000000000000000000000000000',
    hash: '0x0000000000000000000000000000000000000000',
  },
  optionalProps: {
    ensName: 'test',
    ipfsHash: 'Qm',
  },
};
