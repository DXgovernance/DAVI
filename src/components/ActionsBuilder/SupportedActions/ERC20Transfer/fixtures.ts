import { DecodedCall, SupportedAction } from 'components/ActionsBuilder/types';
import ERC20ABI from 'contracts/ERC20.json';
import { BigNumber, utils } from 'ethers';

const ERC20Contract = new utils.Interface(ERC20ABI.abi);

export const erc20TransferEmptyDecodedCallMock: DecodedCall = {
  from: '0x0000000000000000000000000000000000000000',
  callType: SupportedAction.ERC20_TRANSFER,
  function: ERC20Contract.getFunction('transfer'),
  to: '',
  value: BigNumber.from(0),
  args: {
    _to: '',
    _value: BigNumber.from(0),
  },
};

export const erc20TransferDecodedCallMock: DecodedCall = {
  from: '0x0000000000000000000000000000000000000000',
  callType: SupportedAction.ERC20_TRANSFER,
  function: ERC20Contract.getFunction('transfer'),
  to: '0xfFb1cd0F95368DDd06D556161c5D3d9f0f4Fe6d2',
  value: BigNumber.from(0),
  args: {
    _to: '0xaF8eB8C3A5d9d900AA0B98e3Df0bcC17d3C5F698',
    _value: BigNumber.from('5000000000000000000'),
  },
};
