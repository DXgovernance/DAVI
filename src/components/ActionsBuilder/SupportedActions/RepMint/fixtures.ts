import { DecodedCall, SupportedAction } from 'components/ActionsBuilder/types';
import { BigNumber, utils } from 'ethers';
import ERC20SnapshotRep from 'contracts/ERC20SnapshotRep.json';

const ERC20SnapshotRepContract = new utils.Interface(ERC20SnapshotRep.abi);

export const repMintEmptyDecodedCallMock: DecodedCall = {
  from: '0x0000000000000000000000000000000000000000',
  callType: SupportedAction.REP_MINT,
  function: ERC20SnapshotRepContract.getFunction('mint'),
  to: '',
  value: BigNumber.from(0),
  args: {
    to: '',
    amount: BigNumber.from(0),
  },
};

export const repMintDecodedCallMock: DecodedCall = {
  from: '0x0000000000000000000000000000000000000000',
  callType: SupportedAction.REP_MINT,
  function: ERC20SnapshotRepContract.getFunction('mint'),
  to: '0x3f943f38b2fbe1ee5daf0516cecfe4e0f8734351',
  value: BigNumber.from(0),
  args: {
    to: '0x3f943f38b2fbe1ee5daf0516cecfe4e0f8734351',
    amount: BigNumber.from(10000),
  },
};
