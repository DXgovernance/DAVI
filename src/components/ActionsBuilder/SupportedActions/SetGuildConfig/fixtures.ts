import { BigNumber, utils } from 'ethers';
import { SupportedAction } from 'components/ActionsBuilder/types';
import BaseERC20Guild from 'contracts/BaseERC20Guild.json';

export const mockBigNumber = BigNumber.from(100000000);
export const BaseERC20GuildContract = new utils.Interface(BaseERC20Guild.abi);

export const emptyDecodedCall = {
  from: '',
  callType: SupportedAction.SET_GUILD_CONFIG,
  function: BaseERC20GuildContract.getFunction('setConfig'),
  to: '',
  value: BigNumber.from(0),
  args: {
    _proposalTime: '',
    _timeForExecution: '',
    _votingPowerPercentageForProposalExecution: '',
    _votingPowerPercentageForProposalCreation: '',
    _voteGas: '',
    _maxGasPrice: '',
    _maxActiveProposals: '',
    _lockTime: '',
    _minimumMembersForProposalCreation: '',
    _minimumTokensLockedForProposalCreation: '',
  },
  optionalProps: {},
};
