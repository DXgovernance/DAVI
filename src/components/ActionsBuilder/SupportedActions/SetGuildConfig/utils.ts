import { FIELDS } from './constants';
import { BigNumber } from 'ethers';
import { SetGuildConfigFields, ControlField } from './types';

export const bn = (n: string | number | BigNumber) => BigNumber.from(n);

type UpdatedValuesReturn = {
  [key in ControlField]: {
    newValue: BigNumber;
    currentValue: BigNumber;
  };
};
export const getUpdatedValues = (
  current: SetGuildConfigFields,
  modifyed: SetGuildConfigFields
): UpdatedValuesReturn | {} => {
  if (!current || !modifyed) return {};
  return Object.keys(current).reduce((acc, key) => {
    if (!FIELDS.map(f => f.name).includes(key)) return acc;
    const currentValue = bn(current[key] || 0);
    const newValue = bn(modifyed[key] || 0);

    if (!currentValue.eq(newValue)) {
      return {
        ...acc,
        [key]: { newValue, currentValue },
      };
    }
    return acc;
  }, {});
};

export const pickUpdatedOrDefaultValue = (
  current: BigNumber,
  modifyed: BigNumber
): BigNumber => {
  if (BigNumber.isBigNumber(current)) {
    if (BigNumber.isBigNumber(modifyed) && !bn(modifyed).eq(current)) {
      return bn(modifyed);
    }
    return bn(current);
  } else if (BigNumber.isBigNumber(modifyed)) {
    return bn(modifyed);
  }
  return bn(0);
};
