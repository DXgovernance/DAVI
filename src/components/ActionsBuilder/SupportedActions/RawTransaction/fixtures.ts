import { DecodedCall, SupportedAction } from 'components/ActionsBuilder/types';
import { BigNumber } from 'ethers';
import { MOCK_ADDRESS } from 'hooks/Guilds/ens/fixtures';

export const rawDataCallMock: DecodedCall = {
  from: '0x0000000000000000000000000000000000000000',
  callType: SupportedAction.RAW_TRANSACTION,
  function: null,
  to: MOCK_ADDRESS,
  value: BigNumber.from('1000000000000000000'),
  args: null,
  optionalProps: {
    data: '0x01',
  },
};
