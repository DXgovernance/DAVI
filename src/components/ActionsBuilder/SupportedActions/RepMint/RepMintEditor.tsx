import Input from 'components/Primitives/Forms/Input';
import { Button } from 'components/Primitives/Button';
import { Controller, useForm } from 'react-hook-form';
import Avatar from 'components/Avatar';
import React, { useEffect, useState } from 'react';
import { ActionEditorProps } from '..';
import useENSAvatar from 'hooks/Guilds/ether-swr/ens/useENSAvatar';
import { shortenAddress, MAINNET_ID } from 'utils';
import { ReactComponent as Info } from 'assets/images/info.svg';
import useBigNumberToNumber from 'hooks/Guilds/conversions/useBigNumberToNumber';
import { useTotalSupply } from 'hooks/Guilds/guild/useTotalSupply';
import { useTokenData } from 'hooks/Guilds/guild/useTokenData';
import { StyledToolTip } from 'components/ToolTip';
import { useTranslation } from 'react-i18next';
import { ethers } from 'ethers';
import validateRepMint from './validateRepMint';
import {
  Control,
  ControlRow,
  ControlLabel,
} from 'components/Primitives/Forms/Control';
import { Error, StyledInfoIcon, RepMintInput } from './styles';

interface RepMintFormValues {
  repPercent: string;
}
export const Mint: React.FC<ActionEditorProps> = ({
  decodedCall,
  onSubmit,
}) => {
  const { t } = useTranslation();
  const [repAmount, setRepAmount] = useState<string>('0');
  const { parsedData } = useTotalSupply({ decodedCall });
  const { tokenData } = useTokenData();
  const totalSupply = useBigNumberToNumber(tokenData?.totalSupply, 18);
  const { imageUrl } = useENSAvatar(parsedData?.toAddress, MAINNET_ID);

  const { control, handleSubmit, setValue } = useForm<RepMintFormValues>({
    resolver: validateRepMint,
    context: { t },
  });

  useEffect(() => {
    const initialRepAmount = ethers.utils.formatEther(
      parsedData?.amount?.toString()
    );
    if (initialRepAmount) {
      setRepAmount(initialRepAmount);
      setValue(
        'repPercent',
        String((Number(initialRepAmount) * 100) / totalSupply || 0)
      );
    }
  }, []); //eslint-disable-line

  const updateRepAmount = (value: string) => {
    if (!value) {
      setRepAmount('0');
    } else {
      const amount = String((Number(value) / 100) * totalSupply);
      setRepAmount(amount);
    }
  };

  const submitAction = () => {
    onSubmit({
      ...decodedCall,
      args: {
        ...decodedCall.args,
        amount: ethers.utils.parseUnits(repAmount.toString()),
      },
    });
  };

  return (
    <React.Fragment>
      <form onSubmit={handleSubmit(submitAction, console.error)}>
        <Control>
          <ControlLabel>
            {t('repMint.recipient')}
            <StyledInfoIcon src={Info} />
            <StyledToolTip>{t('repMint.recipientTooltip')}</StyledToolTip>
          </ControlLabel>
          <ControlRow>
            <Input
              value={shortenAddress(parsedData?.toAddress)}
              icon={
                <Avatar
                  src={imageUrl}
                  defaultSeed={parsedData?.toAddress}
                  size={18}
                />
              }
              readOnly
              disabled
            />
          </ControlRow>
        </Control>
        <ControlRow>
          <Controller
            name="repPercent"
            control={control}
            render={({ field: { ref, ...field }, fieldState }) => {
              const { invalid, error } = fieldState;

              return (
                <Control>
                  <ControlLabel>
                    {t('repMint.repPercent')}
                    <StyledInfoIcon src={Info} />
                    <StyledToolTip>
                      {t('repMint.repPercentTooltip')}
                    </StyledToolTip>
                  </ControlLabel>
                  <ControlRow>
                    <RepMintInput
                      {...field}
                      onChange={value => {
                        field.onChange(value);
                        updateRepAmount(value);
                      }}
                    />
                  </ControlRow>
                  {invalid && !!error && <Error>{error.message}</Error>}
                </Control>
              );
            }}
          />
        </ControlRow>
        <ControlRow>
          <Control>
            <ControlLabel>
              {t('repMint.repAmount')}
              <StyledInfoIcon src={Info} />
              <StyledToolTip>{t('repMint.repAmountTooltip')}</StyledToolTip>
            </ControlLabel>
            <ControlRow>
              <RepMintInput disabled value={repAmount?.toString()} readOnly />
            </ControlRow>
          </Control>
        </ControlRow>
        <Button m="1rem 0 0" fullWidth type="submit">
          {t('saveAction')}
        </Button>
      </form>
    </React.Fragment>
  );
};

export default Mint;
