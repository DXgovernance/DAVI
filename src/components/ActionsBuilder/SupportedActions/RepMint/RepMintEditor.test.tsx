import { BigNumber } from 'ethers';
import { render } from 'utils/tests';
import { repMintEmptyDecodedCallMock } from './fixtures';
import { Mint } from './RepMintEditor';
const mockBigNumber = BigNumber.from(0);

jest.mock('hooks/Guilds/guild/useTotalSupply', () => ({
  useTotalSupply: () => ({
    parsedData: {
      toAddress: '0x0000000000000000000000000000000000000000',
      amount: mockBigNumber,
    },
  }),
}));

jest.mock('hooks/Guilds/guild/useTokenData', () => ({
  useTokenData: () => ({
    tokenData: {
      symbol: 'REP',
      decimals: 18,
      name: 'Reputation',
      address: '0x0000000000000000000000000000000000000000',
      totalSupply: mockBigNumber,
    },
  }),
}));

jest.mock('hooks/Guilds/ens/useENSAvatar', () => ({
  __esModule: true,
  default: () => ({
    avatarUri: 'test',
    imageUrl: 'test',
    ensName: 'test.eth',
  }),
}));

describe('RepMintEditor', () => {
  it('Should match snapshot', () => {
    const { container } = render(
      <Mint
        decodedCall={repMintEmptyDecodedCallMock}
        updateCall={jest.fn()}
        onSubmit={jest.fn()}
      />
    );
    expect(container).toMatchSnapshot();
  });
});
