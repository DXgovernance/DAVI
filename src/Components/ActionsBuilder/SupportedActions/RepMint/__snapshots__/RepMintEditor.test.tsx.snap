// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepMintEditor renders 1`] = `
<div>
   
  <div
    class="sc-iBPTik sc-crrszt eoNdMM eCejHX"
  >
    <div
      class="sc-iBPTik sc-dQoVA eoNdMM fUOEeB"
    >
      repMint.recipient
      <div
        class="sc-kEjbQP ekqecE sc-hBEYId kmSjMd"
      >
        <svg>
          info.svg
        </svg>
      </div>
      <div
        class="sc-iBPTik sc-iqHYmW eoNdMM gsLXMo"
      >
        repMint.recipientTooltip
      </div>
    </div>
    <div
      class="sc-iBPTik sc-bqyKOL eoNdMM jAxZgR"
    >
      <div
        class="sc-gsTEea fbvUhu"
      >
        <div
          class="sc-dlfnuX jRxgXa"
        >
          <img
            alt="Avatar"
            class="sc-jSgvzq sc-gKseQn iUxzmp fyfiWy"
            data-testid="avatar"
            size="18"
            src="test"
          />
        </div>
        <input
          class="sc-bdfBQB wpJaf"
          readonly=""
          value="0x0000...0000"
        />
        <div
          class="sc-dlfnuX gJuvwh"
        />
      </div>
    </div>
  </div>
  <div
    class="sc-iBPTik sc-bqyKOL eoNdMM jAxZgR"
  >
    <div
      class="sc-iBPTik sc-crrszt eoNdMM eCejHX"
    >
      <div
        class="sc-iBPTik sc-dQoVA eoNdMM fUOEeB"
      >
        repMint.repPercent
        <div
          class="sc-kEjbQP ekqecE sc-hBEYId kmSjMd"
        >
          <svg>
            info.svg
          </svg>
        </div>
        <div
          class="sc-iBPTik sc-iqHYmW eoNdMM gsLXMo"
        >
          repMint.repPercentTooltip
        </div>
      </div>
      <div
        class="sc-iBPTik sc-bqyKOL eoNdMM jAxZgR"
      >
        <div
          class="sc-gsTEea fbvUhu"
        >
          <div
            class="sc-dlfnuX jRxgXa"
          />
          <input
            autocomplete="off"
            autocorrect="off"
            class="sc-bdfBQB wpJaf sc-kstqJO cOlXjq"
            inputmode="decimal"
            maxlength="79"
            minlength="1"
            pattern="^[0-9]*[.,]?[0-9]*$"
            placeholder="0.0"
            spellcheck="false"
            type="text"
            value="0"
          />
          <div
            class="sc-dlfnuX gJuvwh"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    class="sc-iBPTik sc-bqyKOL eoNdMM jAxZgR"
  >
    <div
      class="sc-iBPTik sc-crrszt eoNdMM eCejHX"
    >
      <div
        class="sc-iBPTik sc-dQoVA eoNdMM fUOEeB"
      >
        repMint.repAmount
        <div
          class="sc-kEjbQP ekqecE sc-hBEYId kmSjMd"
        >
          <svg>
            info.svg
          </svg>
        </div>
        <div
          class="sc-iBPTik sc-iqHYmW eoNdMM gsLXMo"
        >
          repMint.repAmountTooltip
        </div>
      </div>
      <div
        class="sc-iBPTik sc-bqyKOL eoNdMM jAxZgR"
      >
        <div
          class="sc-gsTEea fbvUhu"
        >
          <div
            class="sc-dlfnuX jRxgXa"
          />
          <input
            autocomplete="off"
            autocorrect="off"
            class="sc-bdfBQB wpJaf sc-kstqJO cOlXjq"
            inputmode="decimal"
            maxlength="79"
            minlength="1"
            pattern="^[0-9]*[.,]?[0-9]*$"
            placeholder="0.0"
            readonly=""
            spellcheck="false"
            type="text"
            value="0"
          />
          <div
            class="sc-dlfnuX gJuvwh"
          />
        </div>
      </div>
    </div>
  </div>
   
</div>
`;
